### PR TITLE
feat(rag): improve knowledge worker observability

### PR DESCRIPTION
**Pull request by dxian0922**
## Business Background
After the knowledge workers were migrated out of the API service, operators lost timely runtime signals. A task could be received while `progress_msg` remained empty for a long time, and the document, QA import, and GraphRAG workers did not consistently expose startup, queue selection, stage transitions, lock waits, heartbeat, business outcome, and shutdown events.

## Summary
- add a shared structured observability layer for knowledge tasks, including lifecycle, stage, heartbeat, duration, business outcome, and Celery outcome events
- publish `Queued` and intermediate progress for document parsing and QA import before long-running work starts
- instrument document synchronization, Evidence Graph operations, GraphRAG lock waits, retries, skips, partial failures, and terminal outcomes
- validate worker role and selected queues at startup and reduce cold imports for dedicated workers
- add static import-boundary guards and focused observability contract coverage

## Changes
- 22 files changed
- 2,186 insertions and 289 deletions
- scope is limited to `mem-knowledge`

## Risk
- Existing task registration names, queue routing, task arguments, acknowledgement behavior, and retry behavior are preserved.
- Document and QA result semantics are preserved; business outcomes are logged separately from Celery execution outcomes.
- User-visible progress is intentionally populated earlier and updated at additional stages.
- No database migration is required.
- External API Key endpoints under `api/app/controllers/service/rag_api_*_controller.py` are not changed.
- Production database, Elasticsearch, MinerU, model, and upstream successful-business integrations were not exercised in this checkout; the isolated broker/worker validation covers worker lifecycle and failure/skip paths.

## Test Plan
- [x] `PYTHONPATH=mem-knowledge uv run --project mem-knowledge --frozen pytest -q /tmp/memorybear-knowledge-observability-tdd.BL97xE/test_*.py` — 49 passed
- [x] `uv run --project mem-knowledge --frozen ruff check mem-knowledge` — passed
- [x] `uv lock --project mem-knowledge --check` — resolved 144 packages
- [x] `uv run --project mem-knowledge --frozen python mem-knowledge/scripts/check_import_boundaries.py` — passed
- [x] source, wheel, installed-package, and container-rootfs artifact/import scans — passed
- [x] Python 3.12 fresh-environment wheel import and nine-task registry check — passed
- [x] document, GraphRAG, and QA prefork worker startup/shutdown plus role/queue mismatch rejection — passed
- [x] isolated RabbitMQ/Redis execution smoke for document failure, QA failure, GraphRAG skip/failure, lifecycle counts, and graceful shutdown — passed

## Sourcery 摘要

通过结构化生命周期信号、持久化进度更新、阶段埋点以及更安全的 worker 启动验证，提升知识工作者的可观测性。

新功能：
- 为知识工作者任务的生命周期、阶段、进度、心跳、队列耗时、重试和业务结果添加结构化可观测性。
- 暴露文档解析和 QA 导入任务更早的排队状态及中间进度。
- 为 GraphRAG 同步、锁等待、阶段耗时、跳过、重试、合并处理和部分失败添加埋点。

错误修复：
- 将文档和 QA 分发失败显示在持久化进度状态中，避免任务长时间没有及时状态。

增强功能：
- 在启动时根据配置的队列验证 worker 角色，并报告 worker 生命周期事件。
- 减少专用 worker 的冷启动导入，同时保留任务注册、路由、确认和重试行为。
- 传递安全的任务元数据，例如 trace、parent-task、queue 和 source-role 上下文，同时排除敏感的负载字段。

测试：
- 添加针对性的可观测性契约覆盖、导入边界检查、worker 启动验证，以及隔离的 broker/worker 生命周期和失败路径测试。

<details>
<summary>Original summary in English</summary>

## Sourcery 总结

通过结构化任务信号、持久化进度更新、阶段埋点以及更安全的 Worker 启动验证，提升知识工作者的可观测性。

新功能：
- 为文档、QA 导入、同步和 GraphRAG 任务添加结构化的生命周期、阶段、心跳、进度、计时、重试和业务结果可观测性。
- 暴露文档解析和 QA 导入过程中更早的排队状态及中间持久化进度。
- 使用安全的任务元数据报告 GraphRAG 阶段执行、锁等待、跳过、重试、合并处理及部分结果。

Bug 修复：
- 持久化文档和 QA 分发失败状态，避免任务长时间没有及时的状态更新。

增强功能：
- 在启动时根据配置的队列验证 Worker 角色，并发出 Worker 生命周期和 Celery 执行信号。
- 减少专用 Worker 的冷启动导入，同时保留任务注册、路由、确认、参数和重试行为。
- 传递经过清理的追踪信息和父任务上下文，同时防止敏感的任务负载字段进入可观测性日志。

测试：
- 添加可观测性契约测试、导入边界保护、Worker 启动验证，以及隔离的 Broker/Worker 生命周期和失败路径覆盖。

<details>
<summary>Original summary in English</summary>

## Sourcery 总结

通过结构化任务信号、更早持久化进度、详细的阶段级监控，以及更安全的 Worker 启动验证，提升知识工作者任务的可观测性。

新功能：
- 为文档、QA 导入、同步和 GraphRAG 任务添加结构化的生命周期、阶段、心跳、进度、计时、重试和业务结果可观测性。
- 暴露文档解析和 QA 导入任务更早的排队状态及中间持久化进度。
- 在安全的任务上下文中报告 GraphRAG 阶段执行、锁等待、跳过、重试、合并处理和部分结果。

错误修复：
- 持久化文档和 QA 分发失败状态，避免任务长时间没有及时的状态更新。

增强：
- 在启动时根据配置的队列验证 Worker 角色，并报告 Worker 和 Celery 的生命周期事件。
- 减少专用 Worker 冷启动时的导入，同时保留任务注册、路由、确认、参数和重试行为。
- 传递经过清理的追踪信息和父任务元数据，同时从可观测性输出中排除敏感的任务负载字段。

测试：
- 添加可观测性契约测试、导入边界保护测试、Worker 启动验证测试，以及隔离的 Broker/Worker 生命周期和失败路径覆盖。

<details>
<summary>Original summary in English</summary>

## Sourcery 摘要

通过结构化任务信号、持久化进度更新、阶段埋点以及更安全的 Worker 启动验证，提升知识工作者的可观测性。

新功能：
- 为文档、QA 导入、同步和 GraphRAG 任务新增结构化的生命周期、阶段、心跳、进度、耗时、重试和业务结果可观测性。
- 为文档解析和 QA 导入暴露更早的排队状态和中间进度，包括持久化的调度失败信息。
- 在安全的任务上下文中报告 GraphRAG 阶段执行、锁等待、跳过、重试、合并以及部分结果。

Bug 修复：
- 持久化文档和 QA 调度失败信息，使任务状态能够及时更新，而不是一直为空。

增强功能：
- 在启动时根据 Worker 配置的队列验证其角色，并发出 Worker 和 Celery 生命周期信号。
- 减少专用 Worker 冷启动时的导入操作，同时保留任务注册、路由、参数、确认和重试行为。
- 传递经过清理的跟踪信息和父任务元数据，同时将敏感的任务负载字段排除在可观测性输出之外。

测试：
- 新增可观测性契约测试、导入边界保护测试、Worker 启动验证，以及隔离的 Broker/Worker 生命周期和失败路径覆盖。

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Improve knowledge worker observability with structured task signals, persisted progress updates, stage instrumentation, and safer worker startup validation.

New Features:
- Add structured lifecycle, stage, heartbeat, progress, timing, retry, and business-outcome observability for document, QA import, synchronization, and GraphRAG tasks.
- Expose earlier queued and intermediate progress for document parsing and QA imports, including persisted dispatch failures.
- Report GraphRAG stage execution, lock waits, skips, retries, coalescing, and partial outcomes with safe task context.

Bug Fixes:
- Persist document and QA dispatch failures so task status is updated promptly instead of remaining empty.

Enhancements:
- Validate worker roles against their configured queues at startup and emit worker and Celery lifecycle signals.
- Reduce dedicated worker cold-start imports while preserving task registration, routing, arguments, acknowledgements, and retry behavior.
- Propagate sanitized trace and parent-task metadata while excluding sensitive task payload fields from observability output.

Tests:
- Add observability contract tests, import-boundary guards, worker startup validation, and isolated broker/worker lifecycle and failure-path coverage.

</details>

</details>

</details>

</details>

### CODE REVIEW

File: mem-knowledge/src/tasks/observability.py [hunk 1]
```diff
@@ -0,0 +1,717 @@
...
+                document.run = 0
+                if event.business_outcome is BusinessOutcome.SUCCESS:
+                    document.progress = 1.0
+            else:
+                document.run = 1
+                if event.progress is not None:
+                    document.progress = float(event.progress)
+            if messages:
+                document.progress_msg = self._append_message(
+                    document.progress_msg,
```
[the remark this answers is not in the document]
Comment (dxian0922) [reply mem-knowledge/src/tasks/observability.py:300]: 已在 b78787627 按 API 旧服务兼容语义补齐。旧 `api/app/tasks.py` 在取消时直接返回，不会把 progress 设为失败值，且可能遗留 run=1；因此这里不把 ABORTED 映射为 progress=-1。新实现会将 run 持久化为 0、保留取消发生时的 progress，并在 progress_msg 追加 `Task aborted (deleted or cancelled).`。另外，SKIPPED/PARTIAL_FAILURE 当前不经过 DocumentProgressSink。相关契约已纳入本地测试，完整结果为 55 passed。

File: mem-knowledge/src/tasks/evidence_graph.py [hunk 2]
```diff
@@ -93,6 +125,12 @@ def _run_observed(
             countdown,
             int((time.perf_counter() - started_at) * 1000),
         )
+        run.finish(
+            BusinessOutcome.RETRY,
+            error_code="KB_GRAPH_TASK_RETRY",
+            exc=exc,
+            detail=f"retry_countdown_{countdown}s",
+        )
```
[the remark this answers is not in the document]
Comment (dxian0922) [reply mem-knowledge/src/tasks/evidence_graph.py:160]: 已在 4821d6820 修复。现在会在 run.finish 前按 Celery 5.5.3 的规则判断 `request.retries < effective max_retries`：普通 Graph 任务仍为 5 次，文档删除清理仍动态扩展为 8 次；只有确实还有下一次时记录 BusinessOutcome.RETRY，额度耗尽则记录 BusinessOutcome.FAILURE 与 `retries_exhausted`，同时仍调用 task.retry 维持原异常传播。普通 retry、动态 8 次上限和两条耗尽路径均有契约覆盖，完整结果为 55 passed。

**Review by TimeBomb2018 — APPROVED** [review]
